### PR TITLE
Modify Pre/Post-Reconciliation Extensions to provide a Resolver

### LINKED ISSUE #3092: Improve the Prereconciliation extension point
**Describe the current behavior**

Currently the `PreReconciliationChecker` extension point has a `kubeClient.Client` parameter allowing interaction with the cluster; this has proven to be hard to use because we often have a `genruntime.ResourceReference` and `kubeClient.Client` requires a GVK. 

``` go
type PreReconciliationChecker interface {
	PreReconcileCheck(
		ctx context.Context,
		obj genruntime.MetaObject,
		owner genruntime.MetaObject,
		kubeClient kubeclient.Client,
		armClient *genericarmclient.GenericClient,
		log logr.Logger,
		next PreReconcileCheckFunc,
	) (PreReconcileCheckResult, error)
}
```

**Describe the improvement**

Swapping the `kubeClient` parameter for something based on `resolver,Resolver` would still allow access to the cluster, but also would provide the infrastructure to easily handle `genruntime.ResourceReference` when that's all that's available.

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

The two extension points `PreReconciliationChecker` and `PostReconciliationChecker` provide numerous parameters to enable extension implementations to easily achieve their goals.

We passed through a `kubeclient.Client` by design, but have found this to be difficult to consume because our resources often have a `genruntime.ResourceReference` but `kubeclient.Client` requires a GVK.

Replacing `kubeclient.Client` with `resolver.Resolver` bridges this gap, allowing extension implementers to simply call `ResolveReference()` and not fuss with GVK (etc).

Closes #3092 

**How does this PR make you feel**:
gif

### CODE REVIEW

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 1]
```diff
@@ -142,7 +142,7 @@ func (ext *ManagedClusterExtension) PreReconcileCheck(
 	_ context.Context,
 	obj genruntime.MetaObject,
 	_ genruntime.MetaObject,
-	_ kubeclient.Client,
```
Comment (super-harsh) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:145]: Would we not require kubeClient for cluster operations?

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 2]
```diff
@@ -142,7 +142,7 @@ func (ext *ManagedClusterExtension) PreReconcileCheck(
...
+	_ *resolver.Resolver,
```
Comment (super-harsh) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:145]: Consider updating the ADR as well with the change

File: v2/api/containerservice/customizations/managed_cluster_extensions.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/api/containerservice/customizations/managed_cluster_extensions.go:145]: The resolver has a kubeClient already, I think there's a method to access it. If not, easy to add.

File: v2/api/containerservice/customizations/managed_cluster_extensions.go (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/api/containerservice/customizations/managed_cluster_extensions.go:145]: Done. Good catch.

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 3]
```diff
@@ -8,6 +8,7 @@ package customizations
 import (
 	"context"
 	"fmt"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:11]: minor: import ordering

Comment (matthchr) [reply v2/api/containerservice/customizations/managed_cluster_extensions.go:11]: And in other files

**Review by matthchr — APPROVED** [review]

**Review by super-harsh — APPROVED** [review]

File: v2/api/containerservice/customizations/managed_cluster_extensions.go (hunk 3 above)
Comment (theunrepentantgeek) [reply v2/api/containerservice/customizations/managed_cluster_extensions.go]: Fixed.

**Pull request merged**